### PR TITLE
Bookmarks

### LINKED ISSUE #2607: Add support for bookmarking
Allow the user to "book mark" the current state of the application in a manner similar to web browser bookmarks. The user could then quickly toggle between different states. This might be particularly helpful when, for example, editing a colormap. The book mark browser could contain a snapshot of the visualizer window along with optional text

### PR DESCRIPTION
**Pull request by StasJ**
Adds support for bookmarking as described in #2607.

The current snapshot size is the maximum natively supported by MacOS menu bars.

s

s2

Fix #2607

### CODE REVIEW

File: apps/vaporgui/MainForm.cpp [hunk 1]
```diff
@@ -1060,6 +1074,141 @@ void MainForm::_createDeveloperMenu()
...
+    for (auto &b : bookmarks) p->AddBookmark(&b);
+
+    string           activeVizWin = p->GetActiveVizName();
+    ViewpointParams *vpp = _paramsMgr->GetViewpointParams(activeVizWin);
+
+    bool useCustomViewport = vpp->GetValueLong(vpp->UseCustomFramebufferTag, 0);
+    int  customViewportWidth = vpp->GetValueLong(vpp->CustomFramebufferWidthTag, 0);
+    int  customViewportHeight = vpp->GetValueLong(vpp->CustomFramebufferHeightTag, 0);
+
+    int           iconSize = 19;
```
Comment (clyne) [inline apps/vaporgui/MainForm.cpp:1106]: What does this encoded string do? Needs some documentation.

File: include/vapor/Base16StringStream.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <streambuf>
+#include <ostream>
+#include <istream>
+#include <memory>
+
+
+
```
Comment (clyne) [inline include/vapor/Base16StringStream.h:9]: Please document this public class.

File: lib/render/Visualizer.cpp [hunk 3]
```diff
@@ -491,28 +491,39 @@ RegionParams *Visualizer::getActiveRegionParams() const { return _paramsMgr->Get
...
     if (!_getPixelData(framebuffer))
         ;    // goto captureImageEnd;
 
+    if (STLUtils::BeginsWith(path, ":RAM:")) {
+        void *ptr;
+        sscanf(path.c_str(), ":RAM:%p", &ptr);
+
+        memcpy(ptr, framebuffer, sizeof(*framebuffer) * 3 * width * height);
+
+        delete[] framebuffer;
```
Comment (clyne) [inline lib/render/Visualizer.cpp:519]: Any reason to not use a C++ string and avoid the explicit memory management? There is already a goto commented out above that if were uncommented would result in a memory leak.

**Review by clyne — COMMENTED** [review]
Stas, this looks great! Can we make the icons a little bigger? Pretty hard to see. thanks!

File: apps/vaporgui/MainForm.cpp (hunk 1 above)
Comment (StasJ) [reply apps/vaporgui/MainForm.cpp:1106]: Done.

File: include/vapor/Base16StringStream.h (hunk 2 above)
Comment (StasJ) [reply include/vapor/Base16StringStream.h:9]: Done.

File: lib/render/Visualizer.cpp (hunk 3 above)
Comment (StasJ) [reply lib/render/Visualizer.cpp:519]: I don't see how uncommenting the goto statement would result in a memory leak or how using a c++ string would help.

**Comment by StasJ** [comment]
> Stas, this looks great! Can we make the icons a little bigger? Pretty hard to see. thanks!

The current snapshot icon size is the maximum natively supported by MacOS menu bars. The only way around this is to add another non-native menu somewhere else in the GUI which uses custom Qt widgets.

File: lib/render/Visualizer.cpp (hunk 3 above)
Comment (clyne) [reply lib/render/Visualizer.cpp:519]: I said "string" but I should have said C++ vector or array. I.e. a smart pointer that will clean up after itself when it goes out of scope.

Comment (StasJ) [reply lib/render/Visualizer.cpp]: Done.

**Review by clyne — APPROVED** [review]

**Review by shaomeng — APPROVED** [review]

**Review by sgpearse — APPROVED** [review]

**Pull request merged**